pilot/pkg/networking/core/v1alpha3/extension/wasmplugin.go

```diff
@@ -0,0 +1,150 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package extension
+
+import (
+	udpa "github.com/cncf/udpa/go/udpa/type/v1"
+	envoy_config_core_v3 "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	hcm_filter "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/network/http_connection_manager/v3"
+	"github.com/envoyproxy/go-control-plane/pkg/conversion"
+	"google.golang.org/protobuf/proto"
+
+	extensions "istio.io/api/extensions/v1alpha1"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/networking"
+	"istio.io/istio/pilot/pkg/networking/util"
+	authzmodel "istio.io/istio/pilot/pkg/security/authz/model"
+	securitymodel "istio.io/istio/pilot/pkg/security/model"
+)
+
+const (
+	wasmFilterType = "envoy.extensions.filters.http.wasm.v3.Wasm"
+)
+
+var defaultConfigSource = &envoy_config_core_v3.ConfigSource{
+	ConfigSourceSpecifier: &envoy_config_core_v3.ConfigSource_Ads{
+		Ads: &envoy_config_core_v3.AggregatedConfigSource{},
+	},
+	ResourceApiVersion: envoy_config_core_v3.ApiVersion_V3,
+}
+
+// AddWasmPluginsToMutableObjects adds WasmPlugins to HTTP filterChains
+// Note that the slices in the map must already be ordered by plugin
+// priority! This will be the case for maps returned by PushContext.WasmPlugin()
+func AddWasmPluginsToMutableObjects(
+	mutable *networking.MutableObjects,
+	extensionsMap map[extensions.PluginPhase][]*model.WasmPluginWrapper,
+) {
+	if mutable == nil {
+		return
+	}
+
+	for fcIndex, fc := range mutable.FilterChains {
+		// we currently only support HTTP
+		if fc.ListenerProtocol != networking.ListenerProtocolHTTP {
+			continue
+		}
+		mutable.FilterChains[fcIndex].HTTP = injectExtensions(fc.HTTP, extensionsMap)
+	}
+}
+
+func injectExtensions(filterChain []*hcm_filter.HttpFilter, exts map[extensions.PluginPhase][]*model.WasmPluginWrapper) []*hcm_filter.HttpFilter {
+	// copy map as we'll manipulate it in the loop
+	extMap := make(map[extensions.PluginPhase][]*model.WasmPluginWrapper)
+	for phase, list := range exts {
+		extMap[phase] = []*model.WasmPluginWrapper{}
+		extMap[phase] = append(extMap[phase], list...)
+	}
+	newHTTPFilters := make([]*hcm_filter.HttpFilter, 0)
+	for _, httpFilter := range filterChain {
+		switch httpFilter.Name {
+		case securitymodel.EnvoyJwtFilterName:
+			newHTTPFilters = popAppend(newHTTPFilters, extMap, extensions.PluginPhase_AUTHN)
+			newHTTPFilters = append(newHTTPFilters, httpFilter)
+		case securitymodel.AuthnFilterName:
+			newHTTPFilters = popAppend(newHTTPFilters, extMap, extensions.PluginPhase_AUTHN)
+			newHTTPFilters = append(newHTTPFilters, httpFilter)
+		case authzmodel.RBACHTTPFilterName:
```

Comment:
I think this won't work with the CUSTOM and AUDIT action in the authorization policy.

In authorization policy, we have a special CUSTOM action that will add an extra RBAC filter (shadow mode, not used for enforcing but only for matching the request) in the beginning of the filter chain. So the order in the filter chain is:

1. `RBAC filter` shadow mode to match requests for CUSTOM action
2. `ext-authz filter` enforcement of CUSTOM action, triggered based on match results
3. `JWT filter`
4. `istio-authn filter`
5. `RBAC filter` enforcement mode for ALLOW and CUSTOM action

Today, the CUSTOM action does not allow matching on attributes requiring authentication (e.g. mTLS or JWT) because the RBAC shadow mode filter is added before the JWT/authn filter. 

In the future, we will allow users to configure the order of the CUSTOM action. Some users just need it in the beginning of the filter chain (current behavior, useful for use cases like token exchange), other users may choose to put it after authn filter.

The same issue also applies to the AUDIT action where we also reuses the RBAC filter in shadow mode for matching requests instead of doing authorization.

We do not have this WasmPlugin API when we first added the CUSTOM action, it's an internal implementation about which Envoy filter should be used for what purposes, in other words, we do not mandate that jwt filter can only used for authn and RBAC filter can only be used for authz.

The above example of CUSTOM action essentially means authZ may not always happen after the authN. It could be a 2-level authN/Z system where the 1st level is an external authN/Z check with token exchange and the 2nd level is a local authN/Z check on the exchanged token.

In the short term, maybe the API should clarify the definition of the `phase` or add to support for `LOCAL_AUTHN`, `LOCAL_AUTHZ` and `EXTERNAL_AUTH`?

In the long term, I feel this should really align and merge with the ExtensionProvider in the mesh config, maybe something like:

```protobuf
message ExtensionPlugin {
  WorkloadSelector selector = 1;
  PluginPhase phase = 2;
  google.protobuf.Int64Value priority = 3;

  oneof extension_provider {
    WasmPlugin wasm = 4;
    ExternalAuthz ext_authz = 5;
    ZipkinTracing zipkin = 6;
    Stackdriver stackdriver = 7;
    // More extensions to be added.
  }
  // Alternative of using oneof field: repeated ExtensionProvider extension_providers = 4;
}

```

cc @mandarjog @louiscryan @smawson @howardjohn

Reply:
> In the short term, maybe the API should clarify the definition of the phase or add to support for LOCAL_AUTHN, LOCAL_AUTHZ and EXTERNAL_AUTH?

Feels that it might be confusing for Wasm API user to have such detailed phase definition. I would expect most ext authz user won't overlap with Wasm user. We should definitely update the implementation for CUSTOM. For simplicity, maybe we can treat AUTHZ and AUTHN in Wasm API as local ones, and do not consider CUSTOM in Wasm plugin ordering for now (i.e. keep the current behavior to put them at the front of the filter chain), and gather feedback if EXT_AUTH phase is really necessary for Wasm API.

Reply:
okay if we decide to not support ext-authz for now, please make sure to document this properly so that we do not surprise the users.